File: lab2/doublylinkedlist.c
#include<stdio.h>
#include<conio.h>
#include<stdlib.h>
struct DLL{
    int data;
    struct DLL *next,*prev;
};
struct DLL *first,*last=NULL;
void insertion_at_beginning(int element){
    struct DLL *Newnode;
    Newnode=(struct DLL *)malloc(sizeof(struct DLL));
    Newnode->data=element;
    Newnode->next=NULL;
    Newnode->prev=NULL;
    if(first==NULL)
    {
        //No nodes found so our newnode becomes the first, and last at the same time.
        first=last=Newnode;
    }
    else{
        Newnode->next=first;
        first->prev=Newnode;
        first=Newnode;
    }
    printf("%d was inserted . \n",first->data);

}
void insertion_at_end(int element){
    struct DLL *Newnode;
    Newnode=(struct DLL *)malloc(sizeof(struct DLL));
    Newnode->next=NULL;
    Newnode->prev=NULL;
    Newnode->data=element;
    if(first==NULL){
        first=last=Newnode;
    }
    else{
        Newnode->prev=last;
        last->next=Newnode;
        last=Newnode;
    }
    printf("%d has been inserted . \n",last->data);
}
void insertion_at_sp(int element,int pos){
     struct DLL *Newnode;
         int i;
    Newnode=(struct DLL *)malloc(sizeof(struct DLL));
    Newnode->next=NULL;
    Newnode->prev=NULL;
    Newnode->data=element;
    if(pos<=0) printf("Invalid postion.\n");
    else if(pos==1 || first==NULL) insertion_at_beginning(element);
    else{
        struct DLL *temp;
          temp=first;
        for(i=1;i<pos-1 && temp->next!=NULL;i++){
           temp=temp->next;
        }
        if(temp==last){
            insertion_at_end(element);
        }
        else{
            struct DLL *tempP;
            tempP=temp->next;
            temp->next=Newnode;
            Newnode->prev=temp;
            Newnode->next=tempP;
            tempP->prev=Newnode;
        }
        printf("%d was inserted at %d postion.\n",element,pos);
    }
}
void display(){
   struct DLL *temp;
   if(first==NULL){
      printf("List is empty. \n");
   }
   else{
      temp=first;
      while(temp!=NULL){
         printf("%d ->",temp->data);
         temp=temp->next;
      
      }
      printf("NULL \n");
   
   }
}
void del_from_beg(){
    if(first==NULL){
        printf("List is empty ! ");
    }
    else if (first->next==NULL){
        struct DLL *temp;
        temp=first;
        first=last=NULL;
        free(temp);
    }
    else{
        struct DLL *temp;
        temp=first;
        first=first->next;
        printf("%d has been deleted \n",temp->data);
        free(temp);
        first->prev=NULL;
    }
}
void del_from_end(){
      if(first==NULL){
        printf("List is empty ! ");
    }
    else if (first->next==NULL){
       del_from_beg();
    }
    else{
        struct DLL *temp;
        temp=last;
        last=last->prev;
        printf("%d is going to be delted \n",temp->data);
        free(temp);
        last->next=NULL;
      
    }
}
void del_at_sp(int pos){
    int i;
    if(pos<=0 || first==NULL){
        printf("Invalid postion for deletion ! \n");
    }
    else if(pos==1 ||first->next==NULL){
        del_from_beg();
    }
    else{
        struct DLL *temp;
        temp=first;
       for(i=1;i<pos-1 && temp->next!=last;i++){
        temp=temp->next;
       } 
       if(temp->next==last){
        del_from_end();
       }
       else{
        struct DLL *tempP;
        tempP=temp->next;
        temp->next=tempP->next;
        (temp->next)->prev=temp;
        printf("%d is going to delete from the list. \n",tempP->data);
        free(tempP);
       }
    }
}
int main(){
  insertion_at_beginning(100);
  insertion_at_beginning(200);
  insertion_at_end(30);
  insertion_at_sp(40,3);
  display();
  insertion_at_sp(400,1);
  display();
  insertion_at_sp(500,6);
  display();
  del_from_end();
  display();
  del_at_sp(1);
  display();
  del_at_sp(2);
  display();
  return 0;
}

File: lab3assignments/doubly_linked_list.py
class Node:
    def __init__(self,data,next=None,prev=None):
        self.data=data
        self.next=next
        self.prev=prev
class DoublyLinkedList:
    def __init__(self):
        self.head=None

    def insert_at_beginning(self,data): 
            node=Node(data,self.head)
            if self.head is not None:
                self.head.prev=node
            self.head=node
            
    def insert_at_end(self,data):
        if self.head is None:
              self.insert_at_beginning(data)
        else:
            itr=self.head
            while itr.next: 
                itr=itr.next
            itr.next=Node(data,None,itr)
            itr.next.prev=itr
    def insert_at_sp(self,data,pos):
        if pos<=0:
            print("Invalid position for insertion of element !")
            return
        elif pos==1:
             self.insert_at_beginning(data)
        else:
            itr=self.head
            count=1
            while count<pos-1 and itr.next is not None:
                itr=itr.next
                count+=1
            if itr.next==None:
                self.insert_at_end(data)
            else:
                node=Node(data,itr.next,itr)
                itr.next=node
                if node.next is not None:
                  node.next.prev=node
    def del_at_beginning(self):
        if self.head is None:
            print("No elements to perform deletion !")
            return 
        elif self.head.next is None:
            print(f"{self.head.data} is going to be deleted")
            self.head=None
        else :
            print(f"{self.head.data} is going to be deleted! ")
            self.head=self.head.next
            self.head.prev=None
    def del_at_end(self):
        if self.head is None:
            print("No elements to perform deletion !")
            return 
        elif self.head.next is None:
            self.del_at_beginning()
        else:
            itr=self.head
            while itr.next.next is not None:
                itr=itr.next
            print(f"{itr.next.data} is going to be deleted!")
            itr.next.prev=None
            itr.next=None
    def del_at_sp(self,pos):
        if self.head is None:
            print("No elements for deletion !")
            return 
        else: 
            if pos<=0:
                print("Invalid postional argument !")
                return
            elif pos==1:
                self.del_at_beginning()
                return
            itr=self.head
            count=1
            while count<pos-1 and itr.next is not None:
                itr=itr.next
                count+=1
            if itr.next==None:
                self.del_at_end()
            else:
                print(f"{itr.next.data} is going to be delted !") 
                itr.next.next.prev=itr
                itr.next=itr.next.next
                
    def print(self):
        if self.head is None:
            print("List is empty! ")
            return
        itr=self.head
        liststr=''
        while itr:
            liststr+=str(itr.data)+"-->"
            itr=itr.next
        print(liststr)
dll=DoublyLinkedList()
dll.insert_at_beginning(20)
dll.print()
dll.insert_at_beginning(30)
dll.print()
dll.insert_at_end(40)
dll.print()
dll.insert_at_end(50)
dll.print()
dll.insert_at_sp(100,2)
dll.print()
dll.insert_at_sp(200,3)
dll.print()
dll.del_at_beginning()
dll.print()
dll.del_at_end()
dll.print()
dll.del_at_sp(100)
dll.print()
dll.del_at_sp(2)
dll.print()
            
                
        
    
        
            
         